lib/SILOptimizer/Transforms/CSE.cpp

```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
+    SILValue newValue = needsCopy ? copy : newResults[i];
+    if (needsBorrow) {
+      assert(needsCopy);
+      borrow = SILBuilderWithScope(getInsertAfterPoint(newValue).getValue())
+                   .createBeginBorrow(newValue.getLoc(), newValue);
+      newValue = borrow;
+    }
+    while (!oldResults[i]->use_empty()) {
+      Operand *oldUse = *oldResults[i]->use_begin();
+      // Replace with newInstResult if there was no lifetime ending use of
+      // newInstResult. If not, replace with copy of newInstResult.
+      // If there were no opened archetypes defined, we can just replace with
+      // newValue
+      if (!isa<OpenExistentialRefInst>(oldInst)) {
+        oldUse->set(newValue);
+      } else {
+        // Special handling of opened archetypes
+        // Use the cloner to correctly remap the old opened archetype with the
+        // new opened archetype already registered in
+        // CSE::processOpenExistentialRefInst.
+        auto *oldOpenedArchetypeDef = oldUse->getUser();
+        Cloner.getBuilder().setInsertionPoint(oldOpenedArchetypeDef);
+        oldOpenedArchetypeDef->getAllOperands()[oldUse->getOperandNumber()].set(
+            newValue);
+        auto newOpenedArchetypeDef = Cloner.clone(oldOpenedArchetypeDef);
+        oldOpenedArchetypeDef->replaceAllUsesPairwiseWith(
+            newOpenedArchetypeDef);
+        eraseFromParentWithDebugInsts(oldOpenedArchetypeDef);
+      }
+    }
+
+    // We inserted a copy/borrow of the new result, figure out if we need to
+    // insert any destroys/end_borrows of the copy/borrow using
+    // ValueLifetimeAnalysis
+    if (needsCopy || needsBorrow) {
+      SmallVector<SILInstruction *, 4> transitiveUses;
+      if (isa<BeginBorrowInst>(newValue)) {
+        collectGuaranteedTransitiveUses(cast<BeginBorrowInst>(newValue),
+                                        transitiveUses);
+      } else {
+        assert(isa<CopyValueInst>(newValue));
+        std::for_each(newValue->use_begin(), newValue->use_end(),
+                      [&transitiveUses](Operand *op) {
+                        transitiveUses.push_back(op->getUser());
+                      });
+      }
+      ValueLifetimeAnalysis vla(cast<SingleValueInstruction>(newValue),
+                                transitiveUses);
+      ValueLifetimeAnalysis::Frontier frontier;
+      vla.computeFrontier(frontier, ValueLifetimeAnalysis::DontModifyCFG);
+      // Go over all the frontiers and insert a destroy of the copy is a
+      // lifetime ending instruction was not found at the frontier
+      for (auto *end : frontier) {
+        auto endIt = end->getIterator();
+        SILInstruction *lastUser =
+            (endIt == end->getParent()->begin()) ? nullptr : &*std::prev(endIt);
+        bool hasLifetimeEnd = false;
+        // Find if the lastUser of copy was ending its lifetime
+        if (lastUser) {
+          auto lastUserOps = lastUser->getAllOperands();
+          hasLifetimeEnd = std::find_if(
```

Comment:
Two things: first, I'm surprised this doesn't fail to compile because `std::find_if` returns an iterator. I think this should check `std::find_if(...) != lastUserOps.end()`. Better yet, you could use `std::any_of`. Second, you can use `llvm::find_if`, and then you can pass it a container directly instead of a set of iterators (this applies to all algorithms in this patch).

Reply:
yes, thanks! I fixed this instance with std::any_of now.